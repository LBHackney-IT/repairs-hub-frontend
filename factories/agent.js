import { Factory } from 'fishery'

const agentUserFactory = Factory.define(() => ({
  name: 'An Agent',
  email: 'an.agent@hackney.gov.uk',
  roles: ['agent'],
  hasRole: true,
  hasAgentPermissions: true,
  hasContractorPermissions: false,
  hasContractManagerPermissions: false,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const agent = agentUserFactory.build()
