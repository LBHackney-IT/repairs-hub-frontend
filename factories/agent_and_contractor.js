import { Factory } from 'fishery'

export const agentAndContractorUserFactory = Factory.define(() => ({
  name: 'An Operative',
  email: 'an.operative@hackney.gov.uk',
  roles: ['agent', 'contractor'],
}))

export const agentAndContractor = agentAndContractorUserFactory.build()
