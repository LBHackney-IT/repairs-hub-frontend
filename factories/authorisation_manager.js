import { Factory } from 'fishery'

export const authorisationManagerUserFactory = Factory.define(() => ({
  name: 'An Authorisation Manager',
  email: 'a.authorisation_manager@hackney.gov.uk',
  roles: ['authorisation_manager'],
}))

export const authorisationManager = authorisationManagerUserFactory.build()
