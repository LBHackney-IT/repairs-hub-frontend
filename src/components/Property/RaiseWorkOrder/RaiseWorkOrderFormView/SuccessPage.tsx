import {
  createWOLinks,
  LinksWithDRSBooking,
} from '@/root/src/utils/successPageLinks'
import { Property } from '@/root/src/models/propertyTenure'
import { CurrentUser } from '@/root/src/types/variations/types'
import SuccessPage from '@/components/SuccessPage'
import Panel from '../../../Template/Panel'

interface Props {
  immediateOrEmergencyRepairText: boolean
  immediateOrEmergencyDLO: boolean
  authorisationPendingApproval: boolean
  workOrderReference: string
  externallyManagedAppointment: boolean
  property: Property
  currentUser: CurrentUser
  externalAppointmentManagementUrl: string
}

const RaiseWorkOrderSuccessPage = (props: Props) => {
  const {
    immediateOrEmergencyRepairText,
    immediateOrEmergencyDLO,
    authorisationPendingApproval,
    workOrderReference,
    externallyManagedAppointment,
    property,
    currentUser,
    externalAppointmentManagementUrl,
  } = props

  const warningTextToShow = () => {
    if (immediateOrEmergencyRepairText) {
      if (immediateOrEmergencyDLO) {
        return 'Emergency and immediate DLO repairs are sent directly to the planners. An appointment does not need to be booked'
      } else
        return 'Emergency and immediate repairs must be booked immediately. Please call the external contractor'
    } else if (authorisationPendingApproval) {
      return 'Please request authorisation from a manager'
    } else {
      return ''
    }
  }

  return (
    <SuccessPage
      banner={
        <Panel
          title="Work order created"
          authorisationText={
            authorisationPendingApproval && 'but requires authorisation'
          }
          workOrderReference={workOrderReference}
        />
      }
      warningText={warningTextToShow()}
      links={
        externallyManagedAppointment
          ? LinksWithDRSBooking(
              workOrderReference,
              property,
              externalAppointmentManagementUrl,
              currentUser.name
            )
          : createWOLinks(workOrderReference, property)
      }
    />
  )
}

export default RaiseWorkOrderSuccessPage
