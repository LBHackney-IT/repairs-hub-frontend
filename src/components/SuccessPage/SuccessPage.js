import { useContext } from 'react'
import PropTypes from 'prop-types'
import Link from 'next/link'
import WarningText from '../Template/WarningText'
import PageAnnouncement from '../Template/PageAnnouncement'
import { buildDataFromScheduleAppointment } from '../../utils/hact/jobStatusUpdate/notesForm'
import { frontEndApiRequest } from '../../utils/frontEndApiClient/requests'
import UserContext from '../UserContext/UserContext'

const SuccessPage = ({ ...props }) => {
  const { user } = useContext(UserContext)

  const openExternalLinkEventHandler = async () => {
    const jobStatusUpdate = buildDataFromScheduleAppointment(
      props.workOrderReference.toString(),
      `${user.name} opened the DRS Web Booking Manager`
    )

    await frontEndApiRequest({
      method: 'post',
      path: `/api/jobStatusUpdate`,
      requestData: jobStatusUpdate,
    })
  }

  return (
    <>
      <PageAnnouncement
        title={props.text}
        workOrderReference={props.workOrderReference}
      />

      {props.authorisationPendingApproval && (
        <WarningText
          text={`Work order ${props.workOrderReference} requires authorisation. Please request authorisation from a manager.`}
        />
      )}

      <ul className="lbh-list lbh-!-margin-top-9">
        {props.externalSchedulerLink && (
          <li>
            Please{' '}
            <Link href={props.externalSchedulerLink}>
              <a
                className="lbh-link"
                target="_blank"
                rel="noopener"
                onClick={openExternalLinkEventHandler}
              >
                <strong>open DRS</strong>
              </a>
            </Link>{' '}
            to book an appointment
          </li>
        )}

        {props.immediateOrEmergencyDloRepairText && (
          <li>
            <p>
              <strong>
                Emergency and immediate DLO repairs are sent directly to the
                Planners. An appointment does not need to be booked.
              </strong>
            </p>
          </li>
        )}

        {props.workOrderReference && (
          <li>
            <Link href={`/work-orders/${props.workOrderReference}`}>
              <a className="lbh-link">
                <strong>View work order</strong>
              </a>
            </Link>
          </li>
        )}

        {props.shortAddress && (
          <li>
            <Link href={`/properties/${props.propertyReference}`}>
              <a className="lbh-link">
                <strong>Back to {props.shortAddress}</strong>
              </a>
            </Link>
          </li>
        )}

        {props.showSearchLink && (
          <li>
            <Link href="/">
              <a className="lbh-link">
                <strong>Start a new search</strong>
              </a>
            </Link>
          </li>
        )}

        {props.showDashboardLink && (
          <li>
            <Link href="/">
              <a className="lbh-link">
                <strong>Back to dashboard</strong>
              </a>
            </Link>
          </li>
        )}

        {props.showNewWorkOrderLink && (
          <li>
            <Link
              href={`/properties/${props.propertyReference}/raise-repair/new`}
            >
              <a className="lbh-link">
                <strong>Raise a new work order</strong>
              </a>
            </Link>
          </li>
        )}
      </ul>
    </>
  )
}

SuccessPage.propTypes = {
  workOrderReference: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  showDashboardLink: PropTypes.bool,
  showNewWorkOrderLink: PropTypes.bool,
  propertyReference: PropTypes.string,
  shortAddress: PropTypes.string,
  showSearchLink: PropTypes.bool,
  authorisationPendingApproval: PropTypes.bool,
}

export default SuccessPage
