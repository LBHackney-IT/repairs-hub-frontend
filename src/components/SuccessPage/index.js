import PropTypes from 'prop-types'
import Link from 'next/link'
import WarningText from '../Template/WarningText'

const SuccessPage = ({ ...props }) => {
  return (
    <>
      {props.banner}
      {props.warningText && <WarningText text={props.warningText} />}
      <ul className="lbh-list lbh-!-margin-top-9">
        {props.links.map((link, i) => {
          return (
            <li key={i}>
              <Link href={link.href}>
                <a
                  className="lbh-link"
                  onClick={link.onClick}
                  target={link.target}
                  rel={link.rel}
                >
                  {link.text}
                </a>
              </Link>
            </li>
          )
        })}
      </ul>
    </>
  )
}

SuccessPage.propTypes = {
  links: PropTypes.array.isRequired,
  banner: PropTypes.any,
  warningText: PropTypes.string,
}

export default SuccessPage
