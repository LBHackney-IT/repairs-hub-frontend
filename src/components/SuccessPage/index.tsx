import Link from 'next/link'
import WarningText from '../Template/WarningText'

interface Props {
  links: {
    href: string
    text: string
    rel?: string
    target?: string
    onClick?: () => void
  }[]
  banner: any
  warningText?: string
}

const SuccessPage = (props: Props) => {
  const { links, banner, warningText } = props

  return (
    <>
      {banner}
      {warningText && <WarningText text={warningText} />}
      <ul className="lbh-list lbh-!-margin-top-9">
        {links.map((link, i) => {
          return (
            <li key={i}>
              <Link href={link.href}>
                <a
                  className="lbh-link"
                  onClick={link.onClick}
                  target={link.target}
                  rel={link.rel}
                >
                  {link.text}
                </a>
              </Link>
            </li>
          )
        })}
      </ul>
    </>
  )
}

export default SuccessPage
