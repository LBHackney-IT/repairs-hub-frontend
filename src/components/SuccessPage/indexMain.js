// import { useContext } from 'react'
import PropTypes from 'prop-types'
import Link from 'next/link'
import WarningText from '../Template/WarningText'
// import PageAnnouncement from '../Template/PageAnnouncement'
// import { buildDataFromScheduleAppointment } from '@/utils/hact/jobStatusUpdate/notesForm'
// import { frontEndApiRequest } from '@/utils/frontEndApiClient/requests'
// import UserContext from '../UserContext'

const SuccessPage = ({ ...props }) => {
  const { user } = useContext(UserContext)

  //   const openExternalLinkEventHandler = async () => {
  //     const jobStatusUpdate = buildDataFromScheduleAppointment(
  //       props.workOrderReference.toString(),
  //       `${user.name} opened the DRS Web Booking Manager`
  //     )

  //     await frontEndApiRequest({
  //       method: 'post',
  //       path: `/api/jobStatusUpdate`,
  //       requestData: jobStatusUpdate,
  //     })
  //   }

  return (
    <>
      {props.banner}
      {props.warningText && <WarningText text={props.warningText} />}
      <ul className="lbh-list lbh-!-margin-top-9">
        {props.links.map((link, i) => {
          return (
            <li key={i}>
              <Link href={link.href}>
                <a className="lbh-link" onClick={link.onClick}>
                  {link.text}
                </a>
              </Link>
            </li>
          )
        })}
      </ul>
    </>
  )
}

SuccessPage.propTypes = {
  workOrderReference: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  showDashboardLink: PropTypes.bool,
  showNewWorkOrderLink: PropTypes.bool,
  propertyReference: PropTypes.string,
  shortAddress: PropTypes.string,
  showSearchLink: PropTypes.bool,
  authorisationPendingApproval: PropTypes.bool,
  links: PropTypes.array,
  banner: PropTypes.any,
}

export default SuccessPage
