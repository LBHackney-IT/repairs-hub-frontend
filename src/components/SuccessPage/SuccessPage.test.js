import { render } from '@testing-library/react'
import SuccessPage from './SuccessPage'
import { user } from 'factories/agent'
import UserContext from '../UserContext/UserContext'

describe('SuccessPage component', () => {
  describe('Approving / Rejecting a work order', () => {
    const props = {
      workOrderReference: '10000012',
      showDashboardLink: true,
    }

    describe('when variation is approved', () => {
      it('should render success screen with approved message', () => {
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage
              workOrderReference={props.workOrderReference}
              text="You have approved a variation"
              showDashboardLink={props.showDashboardLink}
            />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('when variation is rejected', () => {
      it('should render success screen with rejected message', () => {
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage
              workOrderReference={props.workOrderReference}
              text="You have rejected a variation"
              showDashboardLink={props.showDashboardLink}
            />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })
  })

  describe('Raising a repair', () => {
    const props = {
      workOrderReference: '10000012',
      text: 'Repair work order created',
      propertyReference: '12345678',
      shortAddress: '12 Random Lane',
      showSearchLink: true,
    }

    describe('High cost (over raise limit) authorisation', () => {
      it('should render a success screen with a warning message', () => {
        const testProps = { ...props, authorisationPendingApproval: true }
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage {...testProps} />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('Above target date (over raise limit) authorisation', () => {
      it('should render a success screen with a warning message and link to raise new work order', () => {
        const testProps = {
          ...props,
          authorisationPendingApproval: true,
          showNewWorkOrderLink: true,
        }
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage {...testProps} />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('Within raise limit', () => {
      it('should render a success screen without a warning message', () => {
        const testProps = { ...props, authorisationPendingApproval: false }
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage {...testProps} />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('With an external scheduler URL', () => {
      it('should render text with a link to schedule externally', () => {
        const testProps = {
          ...props,
          externalSchedulerUrl: 'https://drs.example.com',
        }
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage {...testProps} />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('With Emergency and Immediate DLO repairs text', () => {
      it('should render text that Emergency and immediate DLO repairs are sent directly to the Planners', () => {
        const testProps = {
          ...props,
          immediateOrEmergencyDloRepairText: true,
        }
        const { asFragment } = render(
          <UserContext.Provider value={{ user }}>
            <SuccessPage {...testProps} />
          </UserContext.Provider>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })
  })
})
