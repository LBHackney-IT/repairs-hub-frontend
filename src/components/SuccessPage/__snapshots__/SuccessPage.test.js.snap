// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuccessPage component Approving / Rejecting a work order when variation is approved should render success screen with approved message 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <div
      class="lbh-announcement__content"
    >
      <p>
        You have approved a variation for 
        <strong>
          work order 10000012
        </strong>
      </p>
    </div>
  </section>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Back to dashboard
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SuccessPage component Approving / Rejecting a work order when variation is rejected should render success screen with rejected message 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <div
      class="lbh-announcement__content"
    >
      <p>
        You have rejected a variation for 
        <strong>
          work order 10000012
        </strong>
      </p>
    </div>
  </section>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Back to dashboard
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SuccessPage component Raising a repair High cost (over raise limit) authorisation should render a success screen with a warning message 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Repair works order created
    </h3>
    <div
      class="lbh-announcement__content"
    >
      <p>
        Works order number
      </p>
      <strong
        class="govuk-!-font-size-24"
      >
        10000012
      </strong>
    </div>
  </section>
  <div
    class="govuk-warning-text lbh-warning-text"
  >
    <span
      aria-hidden="true"
      class="govuk-warning-text__icon"
    >
      !
    </span>
    <strong
      class="govuk-warning-text__text"
    >
      <span
        class="govuk-warning-text__assistive"
      >
        Warning
      </span>
      Works order 10000012 requires authorisation. Please request authorisation from a manager.
    </strong>
  </div>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/properties/12345678"
      >
        <strong>
          Back to 12 Random Lane
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Start a new search
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SuccessPage component Raising a repair With Emergency and Immediate DLO repairs text should render text that Emergency and immediate DLO repairs are sent directly to the Planners 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Repair works order created
    </h3>
    <div
      class="lbh-announcement__content"
    >
      <p>
        Works order number
      </p>
      <strong
        class="govuk-!-font-size-24"
      >
        10000012
      </strong>
    </div>
  </section>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <p>
        <strong>
          Emergency and immediate DLO repairs are sent directly to the Planners. An appointment does not need to be booked.
        </strong>
      </p>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/properties/12345678"
      >
        <strong>
          Back to 12 Random Lane
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Start a new search
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SuccessPage component Raising a repair With an external scheduler URL should render text with a link to schedule externally 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Repair works order created
    </h3>
    <div
      class="lbh-announcement__content"
    >
      <p>
        Works order number
      </p>
      <strong
        class="govuk-!-font-size-24"
      >
        10000012
      </strong>
    </div>
  </section>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/properties/12345678"
      >
        <strong>
          Back to 12 Random Lane
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Start a new search
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`SuccessPage component Raising a repair Within raise limit should render a success screen without a warning message 1`] = `
<DocumentFragment>
  <section
    class="text-align-center lbh-page-announcement"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Repair works order created
    </h3>
    <div
      class="lbh-announcement__content"
    >
      <p>
        Works order number
      </p>
      <strong
        class="govuk-!-font-size-24"
      >
        10000012
      </strong>
    </div>
  </section>
  <ul
    class="lbh-list lbh-!-margin-top-9"
  >
    <li>
      <a
        class="lbh-link"
        href="/work-orders/10000012"
      >
        <strong>
          View work order
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/properties/12345678"
      >
        <strong>
          Back to 12 Random Lane
        </strong>
      </a>
    </li>
    <li>
      <a
        class="lbh-link"
        href="/"
      >
        <strong>
          Start a new search
        </strong>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
